File: server/ucp-api/src/ports/outbound/project.provider.ts
import { ZitadelRole, ZitadelUser } from '../../domain/dtos/zitadel.dto';

export const PROJECT_PROVIDER = Symbol('PROJECT_PROVIDER');

export interface IProjectProvider {
  createProjectGrant(
    orgId: string,
    projectId: string,
    roleKeys: string[],
  ): Promise<unknown>;
  getRoles(): Promise<ZitadelRole[]>;
  getUsers(orgId: string): Promise<ZitadelUser[]>;
}

File: core/ucp/packages/ucp/src/ucp/ports/outbound/project_provider.py
from typing import Protocol

from ucp.domain.dtos.zitadel_dtos import ZitadelRole, ZitadelUser


class IProjectProvider(Protocol):
    async def create_project_grant(
        self, org_id: str, project_id: str, role_keys: list[str]
    ) -> None:
        """Grants a project to an organization with specific roles"""
        ...

    async def delete_project_grant(self, org_id: str, project_id: str) -> None:
        """Removes a project grant from an organization"""
        ...

    async def get_roles(self) -> list[ZitadelRole]:
        """Gets all roles for the UCP project"""
        ...

    async def get_users(self, org_id: str) -> list[ZitadelUser]:
        """Gets all users with access to this project in a specific org"""
        ...
